Add 4.5-dev5 blog post (#1073)
godotengine/godot-website: collections/_article/dev-snapshot-godot-4-5-dev-5.md

---
title: "Dev snapshot: Godot 4.5 dev 5"
excerpt: The cool chill of the feature freeze approaches…
categories: [pre-release]
author: Thaddeus Crews
image: /storage/blog/covers/dev-snapshot-godot-4-5-dev-5.webp
image_caption_title: Replicube
image_caption_description: A game by Walaber Entertainment LLC
date: 2025-05-30 12:00:00
---

TODO

[Jump to the **Downloads** section](#downloads), and give it a spin right now, or continue reading to learn more about improvements in this release. You can also [try the **Web editor**](https://editor.godotengine.org/releases/4.5.dev5/) or the **Android editor** for this release. If you are interested in the latter, please request to join [our testing group](https://groups.google.com/g/godot-testers) to get access to pre-release builds.

---

> *The cover illustration is from* [**Replicube**](https://www.walaber.com/replicube), *a programming puzzle game where you write code to recreate voxelized objects. It is developed by [Walaber Entertainment LLC](https://www.walaber.com/) ([BlueSky](https://bsky.app/profile/walaber.com), [Twitter](https://twitter.com/walaber)). You can get the game on [Steam](https://store.steampowered.com/app/3401490/Replicube/?curator_clanid=41324400).*

Reviewer: ```suggestion
*The cover illustration is from* [**Replicube**](https://www.walaber.com/replicube), *a programming puzzle game where you write code to recreate voxelized objects. It is developed by [Walaber Entertainment LLC](https://www.walaber.com/) ([Bluesky](https://bsky.app/profile/walaber.com), [Twitter](https://twitter.com/walaber)). You can get the game on [Steam](https://store.steampowered.com/app/3401490/Replicube/?curator_clanid=41324400).*
```

## Highlights

In case you missed them, see the [4.5 dev 1](/article/dev-snapshot-godot-4-5-dev-1/), [4.5 dev 2](/article/dev-snapshot-godot-4-5-dev-2/), [4.5 dev 3](/article/dev-snapshot-godot-4-5-dev-3/), and [4.5 dev 4](/article/dev-snapshot-godot-4-5-dev-4/) release notes for an overview of some key features which were already in those snapshots, and are therefore still available for testing in dev 5.

### Native visionOS support

> Normally, our featured highlights in these development blogs come from longtime contributors. This makes sense of course, as it's generally those users that have the familiarity necessary for major changes/additions that're commonly used for these highlights. That's why it might surprise you to hear that visionOS support comes to us from [Ricardo Sanchez-Saez](https://github.com/rsanchezsaez), whose pull request [GH-105628](https://github.com/godotengine/godot/pull/105628) is their *very first* contribution to the engine! It might *not* surprise you to hear that Ricardo is part of the visionOS engineering team at Apple, which certainly helps get their foot in the door, but that still makes visionOS the first officially-supported platform integration in about a decade.

Reviewer: ```suggestion
Normally, our featured highlights in these development blogs come from long-time contributors. This makes sense of course, as it's generally those users that have the familiarity necessary for major changes or additions that are commonly used for these highlights. That's why it might surprise you to hear that visionOS support comes to us from [Ricardo Sanchez-Saez](https://github.com/rsanchezsaez), whose pull request [GH-105628](https://github.com/godotengine/godot/pull/105628) is his *very first* contribution to the engine! It might *not* surprise you to hear that Ricardo is part of the visionOS engineering team at Apple, which certainly helps get his foot in the door, but that still makes visionOS the first officially-supported platform integration in about a decade.
```
Pronouns taken from https://x.com/rsanchezsaez

> For those unfamiliar, visionOS is Apple's XR environment. We're no strangers to XR as a concept (see our recent [XR blogpost](/article/godot-xr-update-mar-2025/) highlighting the latest [Godot XR Game Jam](https://itch.io/jam/godot-xr-game-jam-feb-2025)), but XR platforms are as unique from one-another as traditional platforms. visionOS users have expressed a strong interest in integrating with our ever-growing XR community, and now we can make that happen. See you all in the next XR Game Jam!

Reviewer: ```suggestion
For those unfamiliar, visionOS is Apple's XR environment. We're no strangers to XR as a concept (see our recent [XR blogpost](/article/godot-xr-update-mar-2025/) highlighting the latest [Godot XR Game Jam](https://itch.io/jam/godot-xr-game-jam-feb-2025)), but XR platforms are as distinct from one another as traditional platforms. visionOS users have expressed a strong interest in integrating with our ever-growing XR community, and now we can make that happen. See you all in the next XR Game Jam!
```

The purpose of an abstract class is to create a baseline for other classes to derive from:
```gdscript
class_name ExtendsMyAbstract extends MyAbstract
```
<img src="/storage/blog/dev-snapshot-godot-4-5-dev-5/abstract-derived.webp" alt="Abstract derived"/>

### Shader baker

From the technical gurus behind implementing [ubershaders](/article/dev-snapshot-godot-4-4-dev-4/#ubershaders-and-pipeline-pre-compilation-and-dedicated-transfer-queues), [Darío Samo](https://github.com/DarioSamo) and [Pedro J. Estébanez](https://github.com/RandomShaper) bring us another miracle of rendering: shader baker exporting! ([GH-102552](https://github.com/godotengine/godot/pull/102552))

<img src="/storage/blog/dev-snapshot-godot-4-5-dev-5/shader-baker.webp" alt="Shader baker"/>

**Before:**
<video autoplay loop muted playsinline>
  <source src="/storage/blog/dev-snapshot-godot-4-5-dev-5/shader-baker-before.webm?1" type="video/webm">
</video>

**After:**
<video autoplay loop muted playsinline>
  <source src="/storage/blog/dev-snapshot-godot-4-5-dev-5/shader-baker-after.webm?1" type="video/webm">
</video>

### Web: WebAssembly SIMD support

> As you might recall, Godot 4.0 initially released under the assumption that multi-threaded web support would become the standard, and only supported that format for web builds. This assumption unfortunately proved to be wishful thinking, and was [reverted in 4.3](/article/dev-snapshot-godot-4-3-dev-3/#single-threaded-web-exports) by allowing for single-threaded builds once more. However, this doesn't mean that these single-threaded environments are inherently incapable of parallel processing; it just requires alternative implementations. One such implementation, <abbr title="Single instruction, multiple data">[SIMD](https://en.wikipedia.org/wiki/Single_instruction,_multiple_data)</abbr>, is a perfect candidate thanks to its support across [most major browsers](https://caniuse.com/wasm-simd). To that end, web-wiz [Adam Scott](https://github.com/adamscott) has taken to integrating this implementation for our web builds by default ([GH-106319](https://github.com/godotengine/godot/pull/106319)).

Reviewer 2: Let's not be ambiguous about dropping support. **All** browsers support SIMD now, in the latest version anyway :)

Also, perhaps it's worth nothing here how massive the performance gains are for Jolt-based physics? @adamscott had two demos in his PR. 

```suggestion
As you might recall, Godot 4.0 initially released under the assumption that multi-threaded web support would become the standard, and only supported that format for web builds. This assumption unfortunately proved to be wishful thinking, and was [reverted in 4.3](/article/dev-snapshot-godot-4-3-dev-3/#single-threaded-web-exports) by allowing for single-threaded builds once more. However, this doesn't mean that these single-threaded environments are inherently incapable of parallel processing; it just requires alternative implementations. One such implementation, <abbr title="Single instruction, multiple data">[SIMD](https://en.wikipedia.org/wiki/Single_instruction,_multiple_data)</abbr>, is a perfect candidate thanks to its support across [all major browsers](https://caniuse.com/wasm-simd). To that end, web-wiz [Adam Scott](https://github.com/adamscott) has taken to integrating this implementation for our web builds by default ([GH-106319](https://github.com/godotengine/godot/pull/106319)).
```

### Inline color pickers

> While it's always been possible to see what kind of variable is assigned to an exported color in the inspector, some users have expressed a keen intererest in allowing for this functionality within the script editor itself. This is because it would mean seeing what kind of color a variable without needing to be exposed, as well as making it more intuitive at a glance as to what color a name/code corresponds to. [Koliur Rahman](https://github.com/dugramen) has blessed us with this QOL goodness, which adds an inline color picker [GH-105724](https://github.com/godotengine/godot/pull/105724). Now no matter where the color is declared, users will be able to immediately and intuitively know what is actually represented in a non-intrusive manner.

Reviewer: ```suggestion
While it's always been possible to see what kind of variable is assigned to an exported color in the inspector, some users have expressed a keen interest in allowing for this functionality within the script editor itself. This is because it would mean seeing what kind of color is represented by a variable without it needing to be exposed, as well as making it more intuitive at a glance as to what color a name or code corresponds to. [Koliur Rahman](https://github.com/dugramen) has blessed us with this QOL goodness, which adds an inline color picker [GH-105724](https://github.com/godotengine/godot/pull/105724). Now no matter where the color is declared, users will be able to immediately and intuitively know what is actually represented in a non-intrusive manner.
```

> While it's always been possible to see what kind of variable is assigned to an exported color in the inspector, some users have expressed a keen intererest in allowing for this functionality within the script editor itself. This is because it would mean seeing what kind of color a variable without needing to be exposed, as well as making it more intuitive at a glance as to what color a name/code corresponds to. [Koliur Rahman](https://github.com/dugramen) has blessed us with this QOL goodness, which adds an inline color picker [GH-105724](https://github.com/godotengine/godot/pull/105724). Now no matter where the color is declared, users will be able to immediately and intuitively know what is actually represented in a non-intrusive manner.

Reviewer 3: ```suggestion
While it's always been possible to see what kind of variable is assigned to an exported color in the inspector, some users have expressed a keen intererest in allowing for this functionality within the script editor itself. This is because it would mean seeing what kind of color a variable without needing to be exposed, as well as making it more intuitive at a glance as to what color a name/code corresponds to. [Koliur Rahman](https://github.com/dugramen) has blessed us with this quality-of-life goodness, which adds an inline color picker [GH-105724](https://github.com/godotengine/godot/pull/105724). Now no matter where the color is declared, users will be able to immediately and intuitively know what is actually represented in a non-intrusive manner.
```

## Changelog

**109 contributors** submitted **252 fixes** for this release. See our [**interactive changelog**](https://godotengine.github.io/godot-interactive-changelog/#4.5-dev5) for the complete list of changes since the previous 4.5-dev4 snapshot.

This release is built from commit [`64b09905c`](https://github.com/godotengine/godot/commit/64b09905c7b2877f8aef99d8b63e73e5d31acfb9).

## Downloads

{% include articles/download_card.html version="4.5" release="dev5" article=page %}

**Standard build** includes support for GDScript and GDExtension.

**.NET build** (marked as `mono`) includes support for C#, as well as GDScript and GDExtension.

{% include articles/prerelease_notice.html %}

## Known issues

- Windows executables (both the editor and export templates) have been signed with an expired certificate. You may see warnings from Windows Defender's SmartScreen when running this version, or outright be prevented from running the executables with a double-click ([GH-106373](https://github.com/godotengine/godot/issues/106373)). Running Godot from the command line can circumvent this. We will soon have a renewed certificate which will be used for future builds.

> - `get_class_icon()` can erroneously ignore its fallback argument, causing a crash. A fix is [already in place](https://github.com/godotengine/godot/pull/106963), but will arrive in the following snapshot.

Reviewer: ```suggestion
- `get_class_icon()` can erroneously ignore its fallback argument, causing a crash. A fix is [on its way](https://github.com/godotengine/godot/pull/106963), and will arrive in the following snapshot.
```

Reviewer 4: btw `get_class_icon()` is not an exposed method, so this note does not make much sense like this.

Author: I'll remove the reference entirely for the time being. If there's a sensible way of framing this, we can include that instead

<img src="/storage/blog/dev-snapshot-godot-4-5-dev-5/shader-baker.webp" alt="Shader baker"/>

**Before:**
<video autoplay loop muted playsinline>
  <source src="/storage/blog/dev-snapshot-godot-4-5-dev-5/shader-baker-before.webm?1" type="video/webm">
</video>

**After:**
<video autoplay loop muted playsinline>
  <source src="/storage/blog/dev-snapshot-godot-4-5-dev-5/shader-baker-after.webm?1" type="video/webm">
</video>

However, it comes with tradeoffs:

1. Export time will be much longer.
2. Build size will be much larger since the baked shaders can take up a lot of space.

> 3. We have removed several MoltenVK bug workarounds from the Forward+ shader, therefore we no longer guarantee support for the Forward+ renderer on intel Macs. If you are targeting intel Macs, you should use the Mobile or Compatibility renderers.

Reviewer: ```suggestion
3. We have removed several MoltenVK bug workarounds from the Forward+ shader, therefore we no longer guarantee support for the Forward+ renderer on Intel Macs. If you are targeting Intel Macs, you should use the Mobile or Compatibility renderers.
```

### Inline color pickers

While it's always been possible to see what kind of variable is assigned to an exported color in the inspector, some users have expressed a keen interest in allowing for this functionality within the script editor itself. This is because it would mean seeing what kind of color is represented by a variable without it needing to be exposed, as well as making it more intuitive at a glance as to what color a name or code corresponds to. [Koliur Rahman](https://github.com/dugramen) has blessed us with this quality-of-life goodness, which adds an inline color picker [GH-105724](https://github.com/godotengine/godot/pull/105724). Now no matter where the color is declared, users will be able to immediately and intuitively know what is actually represented in a non-intrusive manner.

<img src="/storage/blog/dev-snapshot-godot-4-5-dev-5/inline-color-picker.webp" alt="Inline color picker"/>

### Rendering goodies

> The renderer got a fair amount of love this snapshot; not from any one PR, but rather a multitude of community members bringing some long-awaited features to light. [Raymond DiDonato](https://github.com/RGDTAB) helped SMAA 1x make its transition from addon to fully-fledged engine feature ([GH-102330](https://github.com/godotengine/godot/pull/102330)). [Capry](https://github.com/LunaCapra) brings bent normal maps to further enhance specular occlusion and indirect lighting ([GH-89988](https://github.com/godotengine/godot/pull/89988)). Our very own [Clay John](https://github.com/clayjohn) converted our Compatibility backend to use a fragment shader copy instead of a blit copy, working around common sample rate issues on mobile devices ([GH-106267](https://github.com/godotengine/godot/pull/106267)). More technical information on these rendering changes can be found in their associated PRs.

Reviewer 5: These (SMAA and bent normals) should be accompanied by before and after images. There are some very nice examples in the relevant PRs. 

Especially bent normals which is very impressive

<img src="/storage/blog/dev-snapshot-godot-4-5-dev-5/shader-baker.webp" alt="Shader baker"/>

**Before:**
<video autoplay loop muted playsinline>
  <source src="/storage/blog/dev-snapshot-godot-4-5-dev-5/shader-baker-before.webm?1" type="video/webm">
</video>

**After:**
<video autoplay loop muted playsinline>
  <source src="/storage/blog/dev-snapshot-godot-4-5-dev-5/shader-baker-after.webm?1" type="video/webm">
</video>

However, it comes with tradeoffs:

1. Export time will be much longer.
2. Build size will be much larger since the baked shaders can take up a lot of space.
3. We have removed several MoltenVK bug workarounds from the Forward+ shader, therefore we no longer guarantee support for the Forward+ renderer on Intel Macs. If you are targeting Intel Macs, you should use the Mobile or Compatibility renderers.

> 4. Baking for Vulkan can be done from any device, but baking for D3D12 needs to be done from a Windows device and baking for Apple devices need to be from an Apple device.

Reviewer 5: Please see the comments from Stuart about this, he provided a lot of important clarification in response to my questions

https://github.com/godotengine/godot/pull/102552#issuecomment-2917363298
